Update dependencies on the word converter

BillWagner opened the pull request:
Fixes #997 

This PR updates all the NuGet packages for the word converter, and the other tools.

The first commit updates every NuGet package to the "latest stable" except the FSharp.Formatting package.

After that, I updated the FSharp.Formatting package iteratively in successive commits to get it to the latest version. In most cases, I fixed issues as they appeared. Many were compilation errors because new arguments were added to methods, or namespaces changed.

v 16.0.1 fixed the issue Jon had noted regarding parsing emphasis blocks. In that update, I removed the block of code that worked around that issue.

That same update introduced some false positive unit test failures. The new version created text nodes with a single space character before changing styles. The appearance was the same. Even after updating the existing results these tests still failed. *This is the first area I'd like help fixing.* The validation code seems to view those nodes as empty, not having a single space. It also introduced one new bug.  If a block of emphasis text didn't end with at the end of a word, the parser didn't find the end of the emphasis block, but rather added the `*` character in the text and extended the emphasis text until the next end marker was found. I added a new test for this. It's the test resource for "emphasis-with-plural.md".

v 16.0.2 fixed a parsing error introduced in the previous update. However, the unit test still failed with the same false positive noted above. In addition, it introduced a new failure. Starting in this version, tables that included bold code were are incorrect.

The source is `**+x**`, with the "+x" also code font:  **`x`**. The output is italicized, with the additional `*`:  `*+x*`. I added a new test for this, and tried to update the XML output correctly.

Those failures remain through the additional updates until the current version.

BillWagner committed:
Update Most packages

Update all packages except FSharp.Formatting. That oner has breaking changes.

BillWagner committed:
update to 5.02

Fix compiler errors in 5.0.2

Run all tests, and smoke test the Word doc output.

BillWagner committed:
Update to V6.02

BillWagner committed:
success in v7.1

Upgrade to v7.1, ran tests and converter.

BillWagner committed:
Conversion to 10.1.1

Versions 8 and 9 were compatible.

Version 10 introduced breaking changes in the InlineBlock -> InlineHTMLBlock.

BillWagner committed:
Update to 13.0.1

Version 13 had a breaking change with an additional optional parameter.

BillWagner committed:
V14.01 is fine

Update to V 14.0.1 succeeds.

BillWagner committed:
V15.0.3 works fine

BillWagner committed:
Checkin at 16.0.1

This version fixes some parsing bugs, so a workaround has been removed.

However, it introduces a markdown parser bug where *some_grammar_production*s (note trailing 's') doesn't properly close the emphasis.

I've written a new failing test, and I'm going to continue to update to the latest version, so see if it's fixed in a later version before fixing this failing test.

BillWagner committed:
v 16.02: some failures

This update fixed one issue, and introduced a new one. It also caused some breaking tests.

Fix: Emphasis nodes like `*grammar_productions*s` now parse correctly. However, the test I introduced doesn't pass. (More below).

New breaking change: The tables where the first column is **bold** is parsed as emphasis, with `*` surrounding each element.

New test failures: Three tests, and the one I added for emphasis parsing fail, but are false failures. The differences in the XML are the nodes with a single unformatted space character.

Moving forward, to see if any of the issues are fixed going forward.

BillWagner committed:
Update to v 17

No change in behavior.

Now, the warning on dependencies has gone away. (The F# formatter approves of the new F# compiler)

BillWagner committed:
No change, up to latest release

Now at v 19.1.1, which is the latest stable.

Same behavior.

BillWagner commented:
@jskeet 

This isn't ready to merge, but it's worth taking a look.

Any pointers on fixing the false positive unit tests is appreciated. I'll work on trying to fix the actual issue regarding the math operator tables.

BillWagner committed:
Fixed the tables.

BillWagner commented:
Marking this ready, as I now need help on getting the unit tests to pass.

BillWagner commented on tools/MarkdownConverter.Tests/markdown-lint.xml:
```diff
@@ -20,7 +20,10 @@
     </r>
     <r>
-      <t space="preserve">: Note 1. </t>
+      <t space="preserve">: Note 1.</t>
+    </r>
+    <r>
+      <t space="preserve"> </t>
```
These are the diffs in output that are causing the unit tests to fail. The updated XML does match the generated output, but for some reason, the difference parser is viewing the new nodes as "null text" rather than a single space.

BillWagner replied:
@jskeet 

In my final two commits, I fix the unit tests, by modifying the comparison tests. Only add a failed comparison when the expect node is null and the actual node is a single space.

BillWagner commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -662,33 +682,6 @@ IEnumerable<OpenXmlElement> Span2Elements(MarkdownSpan md, bool nestedSpan = fal
         {
             IEnumerable<MarkdownSpan> spans = (md.IsStrong ? ((MarkdownSpan.Strong) md).body : ((MarkdownSpan.Emphasis) md).body);
 
-            // Workaround for https://github.com/tpetricek/FSharp.formatting/issues/389 - the markdown parser
```
This issue has been fixed, so this workaround has been removed.

BillWagner commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -951,7 +944,9 @@ IEnumerable<OpenXmlElement> Literal2Elements(string literal, bool isNested, bool
     }
 
-    IEnumerable<OpenXmlCompositeElement> GenerateCustomBlockElements(string customBlockId, MarkdownParagraph.InlineBlock block) => customBlockId switch
+    // Note for future PR: "function_members", "format_strings_1", and "format_strings_2" are the
+    // only special cases still used. The others can be safely removed in a future PR.
+    IEnumerable<OpenXmlCompositeElement> GenerateCustomBlockElements(string customBlockId, MarkdownParagraph.InlineHtmlBlock block) => customBlockId switch
```
@jskeet Let me know if you want me to remove the (now) unused custom blocks in this PR.

jskeet replied:
Looking into it now - I need to understand the comment a bit more first.

BillWagner committed:
update test

BillWagner committed:
Well, make the first row bold

Now that the first column in these tables is bold, make the first row (in the markdown) bold as well.

BillWagner commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -424,6 +423,26 @@ IEnumerable<OpenXmlCompositeElement> Paragraph2Paragraphs(MarkdownParagraph md)
                 {
                     var mdcell = mdrow[icol];
                     var cell = new TableCell();
+
```
I freely admit that this is ugly.

But it works.

Any suggestions welcome.  I did try a pattern match and that was less readable.

jskeet commented on tools/MarkdownConverter.Tests/MarkdownConverter.Tests.csproj:
```diff
@@ -11,6 +11,11 @@
     <EmbeddedResource Include="*.md" />
   </ItemGroup>
+  
+  <ItemGroup>
```
I suspect this block isn't necessary/desirable.

jskeet commented on tools/MarkdownConverter.Tests/emphasis-with-plural.md:
```diff
@@ -0,0 +1,6 @@
+
+Rules:
+
+- If `x` is zero and the namespace declaration contains an *extern_alias_directive* or *using_alias_directive* that associates the name�`I` with an imported namespace or type, then the *namespace_or_type_name* refers to that namespace or type.
```
I'm seeing odd non-ASCII characters here after "name". Are those deliberate?

jskeet commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -423,6 +423,27 @@ IEnumerable<OpenXmlCompositeElement> Paragraph2Paragraphs(MarkdownParagraph md)
                     var cell = new TableCell();
+
+                    // This logic deals with the special case where
+                    // the cell is meant to cntain bold inline code text.
```
Nit: cntain => contain

jskeet commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -423,6 +423,27 @@ IEnumerable<OpenXmlCompositeElement> Paragraph2Paragraphs(MarkdownParagraph md)
+                    var span2 = para?.body.Skip(1).FirstOrDefault() as MarkdownSpan.Emphasis;
+                    var cellText = span2?.body.FirstOrDefault() as MarkdownSpan.InlineCode;
+                    var trailingAsterisk = span2?.body.Skip(1).FirstOrDefault() as MarkdownSpan.Literal;
+                    if (leadingAsterisk?.text == "*" &&  cellText != null && trailingAsterisk?.text == "*")
```
Nit: double space before cellText

jskeet commented on tools/MarkdownConverter/Converter/MarkdownSourceConverter.cs:
```diff
@@ -923,6 +944,8 @@ IEnumerable<OpenXmlElement> Literal2Elements(string literal, bool isNested, bool
         }
     }
 
+    // Note for future PR: "function_members", "format_strings_1", and "format_strings_2" are the
```
I don't understand this comment. What's happened to CreateMultiplicationTable etc? (Did we stop *using* those a while ago, but didn't remove the code?)

jskeet replied:
Right - I see now that the tables were removed in https://github.com/dotnet/csharpstandard/pull/584.

I suggest leaving the code here for now, but I've filed https://github.com/dotnet/csharpstandard/issues/1046 to remove it in a separate PR. I'm happy to do that as soon as this one is merged.

BillWagner replied:
Added a comment to make it easier to remember this

jskeet reviewed (approved):
I'm looking at the custom block bit now. While the workaround for the bold in tables is a bit complicated, my experience is that it's almost always really hard to make code like this any more readable.

BillWagner committed:
Respond to feedback

Respond to doe review feedback.

BillWagner committed:
Tests pass

I added logic in the comparison lambda to ignore the special cases where the target is either the empty string, or null, and the test details is a single space.

BillWagner committed:
Refactor the check

Refactor the check for comparing a null element with a single space text element.
